File: OOP/6_multilevel_inheritance.cpp
#include <iostream>
#include <string>
using namespace std;


class Grandparent {
protected:
    string grandparent_name;

public:

    Grandparent(string grandparent_name) {
        this->grandparent_name = grandparent_name;
    }


    void displayGrandparent() {
        cout << "Grandparent Name: " << grandparent_name << endl;
    }
};


class Parent : public Grandparent {
protected:
    string parent_name;

public:
   
    Parent(string grandparent_name, string parent_name) 
        : Grandparent(grandparent_name) {
        this->parent_name = parent_name;
    }


    void displayParent() {
        cout << "Parent Name: " << parent_name << endl;
    }
};


class Child : public Parent {
private:
    string child_name;

public:

    Child(string grandparent_name, string parent_name, string child_name) 
        : Parent(grandparent_name, parent_name) {
        this->child_name = child_name;
    }


    void displayChild() {
        cout << "Child Name: " << child_name << endl;
    }
};

int main() {

    Child child("John Sr.", "John Jr.", "Johnny");

 
    child.displayGrandparent();  
    child.displayParent();       
    child.displayChild();       

    return 0;
}

File: OOP/6_multilevel_inheritance.py
# Base class (grandparent class)
class Grandparent:
    def __init__(self, grandparent_name):
        self.grandparent_name = grandparent_name

    def display_grandparent(self):
        print(f"Grandparent Name: {self.grandparent_name}")


# Intermediate class (parent class), inherits from Grandparent
class Parent(Grandparent):
    def __init__(self, grandparent_name, parent_name):
        # Call the constructor of Grandparent
        super().__init__(grandparent_name)
        self.parent_name = parent_name

    def display_parent(self):
        print(f"Parent Name: {self.parent_name}")


# Derived class (child class), inherits from Parent
class Child(Parent):
    def __init__(self, grandparent_name, parent_name, child_name):
        # Call the constructor of Parent (which also calls Grandparent)
        super().__init__(grandparent_name, parent_name)
        self.child_name = child_name

    def display_child(self):
        print(f"Child Name: {self.child_name}")


child = Child("John Sr.", "John Jr.", "Johnny")

# Accessing methods from all levels of inheritance
child.display_grandparent()  
child.display_parent()       
child.display_child()        
